File: July13/GALACTIK.cpp
#include<iostream>
#include<algorithm>
#include<cstdio>
#include<vector>
#include<map>
using namespace std;
typedef struct node{
	int x;
	int y;
};
bool myfunc(struct node s1,struct node s2){
	return ((s1.y<s2.y)?true:false);
}
class Helper{
	private:
		int *arr;
		int size;
		int paths;
		int swap;
		map<int,int> Data;
		map<int,int>::iterator it,it1;
	public:
		Helper(int N,int M){
			size=N+1;
			paths=M;
			arr=new int[size];
			int i=0;
			while(i<size){
				arr[i]=i;
				i++;
			}
		}
		void setStructure(){
			int i=1;
			int A,B;
			while(i<=paths){
				scanf("%d %d",&A,&B);
				if(A>B){
					unionData(B,A);
				}
				else{
					unionData(A,B);
				}
				i++;
			}
		}
		void unionData(int x,int y){
			x=find(x);
			y=find(y);
			if(x!=y){
				if(x<y){
					arr[y]=x;
				}
				else{
					arr[x]=y;
				}
			}
		}
		int find(int x){
			while(x!=arr[x]){
				x=arr[x];
			}
			return x;
		}
		void setWeights(){
			int i=1;
			int C;
			int X;
			while(i<=size){
				scanf("%d",&C);
				if(arr[i]==i){
					Data[i]=C;
				}
				else{
					X=find(i);
					it=Data.find(X);
					if((*it).second>=0){
						if(C>=0 && C< ((*it).second)){
							(*it).second=C;
						}
					}
					else{
						if(C>=0){
							(*it).second=C;
						}
					}
				}
				i++;
			}
		}
		long getCost(){
			if(Data.size()==1){
				return 0;
			}
			long Total=0;
			vector<node> arr;
			it=Data.begin();
			while(it!=Data.end()){
				if((*it).second<0){
					return -1;
				}
				node Node;
				Node.x=(*it).first;
				Node.y=(*it).second;
				arr.push_back(Node);
				it++;
			}
			sort(arr.begin(),arr.end(),myfunc);
			int i=0;
			int j=i+1;
			while(j<arr.size()){
				Total+=arr[i].y+arr[j].y;
				j++;
			}
			return Total;
		}
};
int main(){
	int N,M;
	int swap;
	scanf("%d %d",&N,&M);
	if(N==1){
		printf("0\n");
	}
	else{
		Helper ob(N,M);
		ob.setStructure();
		ob.setWeights();
		printf("%ld\n",ob.getCost());
	}
	return 0;
}

File: July13/GALACTIK.java
import java.io.*;
import java.util.*;
class GALACTIK{
	private static int arr[];
	private static int N,M;
	private static int j;
	static Map<Integer,Integer> Data;
	private static void union(int x,int y){
		x=find(x);
		y=find(y);
		if(x<y){
			arr[y]=x;
		}
		else{
			arr[x]=y;
		}
	}
	private static int find(int x){
		while(x!=arr[x]){
			x=arr[x];
		}
		return x;
	}
	private static void setStructure(BufferedReader I)throws IOException{
		int i=1;
		String arr[];
		int x,y;
		while(i<=M){
			arr=(I.readLine()).split(" ");
			x=Integer.parseInt(arr[0]);
			y=Integer.parseInt(arr[1]);
			if(x<y){
				union(x,y);
			}
			else{
				union(y,x);
			}
			i++;
		}
	}
	private static void setCost(BufferedReader I)throws IOException{
		int i=1;
		int Cost;
		int x;
		int temp;
		while(i<=N){
			Cost=Integer.parseInt(I.readLine());
			if(i==arr[i]){
				Data.put(i,Cost);
			}
			else{
				x=find(i);
				temp=Data.get(x);
				if(Cost>0 && temp>Cost){
					Data.put(x,Cost);
				}
			}
			i++;
		}
	}
	private static long getCost(){
		Iterator it=Data.entrySet().iterator();
		Iterator it1;
		int key1,key2,value1,value2;
		long Total=0;
		while(it.hasNext()){
			Map.Entry pairs=(Map.Entry)it.next();
			key1=(Integer)pairs.getKey();
			value1=(Integer)pairs.getValue();
			if(value1<0){
				return -1;
			}
			it1=it;
			if(it1.hasNext()){
				Map.Entry pairs2=(Map.Entry)it1.next();
				key2=(Integer)pairs2.getKey();
				value2=(Integer)pairs2.getValue();
				if(value2<0){
					return -1;
				}
				System.out.println(key1+" "+value1+" "+key2+" "+value2);
				if(value2<value1){
					Data.put(key1,value2);
				}
				Total+=value1+value2;
				Data.remove(key2);
			}
			else{
				break;
			}
		}
		return Total;
	}
	public static void main(String args[])throws IOException{
		BufferedReader I=new BufferedReader(new InputStreamReader(System.in));
		Data=new HashMap<Integer,Integer>();
		String arr1[]=(I.readLine()).split(" ");
		N=Integer.parseInt(arr1[0]);
		M=Integer.parseInt(arr1[1]);
		arr=new int[N+1];
		j=0;
		while(j<=N){
			arr[j]=j;
			j++;
		}
		setStructure(I);
		setCost(I);
		System.out.println(getCost());
	}
}
